pkg/controller/utils/utils.go

```diff
@@ -174,6 +175,41 @@ func AddSecretWatchWithLabel(c controller.Controller, ns, label string) error {
 	})
 }
 
+// AddCSRWatchWithLabel adds a watch for CSRs with the given label.
+func AddCSRWatchWithLabel(c controller.Controller, labelName string) error {
+	return c.Watch(&source.Kind{Type: &certificatesv1.CertificateSigningRequest{}}, &handler.EnqueueRequestForObject{}, &predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			_, hasLabel := e.Object.GetLabels()[labelName]
+			return hasLabel
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			_, hasLabel := e.ObjectNew.GetLabels()[labelName]
+			if !hasLabel {
+				return false
+			}
+			csr, ok := e.ObjectNew.(*certificatesv1.CertificateSigningRequest)
+			if !ok {
+				return false
+			}
+			if csr.Spec.SignerName != OperatorCSRSignerName {
+				return false
+			}
+			for _, condition := range csr.Status.Conditions {
+				if (condition.Type == certificatesv1.CertificateDenied || condition.Type == certificatesv1.CertificateFailed) && condition.Status == corev1.ConditionTrue {
+					// These request statuses are deterministic/non-recoverable.
+					return false
+				}
+			}
+			// We are only interested in CSRs that have not been signed. This also excludes
+			// us from watching our own updates and reconciling for no reason.
+			return csr.Status.Certificate == nil
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return false // We are not interested in these events.
```

Comment:
For future readers, it might be worth elaborating on why we're not interested. From experience, best not to assume the reader knows why. In other words, "We are not interested in these events" is self-evident by the fact that we return "false" here unconditionally. The interesting thing is "why?"